lib/puppet/data_providers/hocon_data_provider_factory.rb

```diff
@@ -0,0 +1,28 @@
+# This file is loaded by the autoloader, and it does not find the data function support unless required relative
+#
+require 'hocon'
+require 'hocon/config_error'
+require 'hocon/config_syntax'
+
+module Puppet::DataProviders
+  class HoconDataProviderFactory < Puppet::Plugins::DataProviders::FileBasedDataProviderFactory
+    def create(name, paths, parent_data_provider)
+      HoconDataProvider.new(name, paths, parent_data_provider)
+    end
+
+    def path_extension
+      '.hocon'
+    end
+  end
+
+  class HoconDataProvider < Puppet::Plugins::DataProviders::PathBasedDataProvider
+    def initialize_data(path, lookup_invocation)
+      data = Hocon.load(File.absolute_path(path), {:syntax => Hocon::ConfigSyntax::HOCON})
+      HieraConfig.symkeys_to_string(data.nil? ? {} : data)
```

Comment:
Will the Hocon parser produce keys that are ruby Symbols? The Hiera ruby implementation has a special interpretation for keys starting with ':' but that's not standard. This line should be removed unless the Hocon implementation has the same behavior,